### PR TITLE
Improve CLI scanner detection/refusal of invalid hostnames

### LINKED ISSUE #192: Refuse hostnames that end in a port number
People are trying to scan specific ports using, I *guess*, ```scan('hostname:80')``` or ```scan('hostname:443')```, which is having bad results (#180, #191), as everything internally expects a hostname, not a host:port.

Consider prohibiting hostnames from ending in :port entirely, since as noted in #191, the scanner will always scan :80 and :443 due to redirect tests.

### PR DESCRIPTION
**Pull request by floatingatoll**
The CLI scanner did not refuse hostnames on sight for containing a protocol prefix or port suffix. This adds a simple test to reject these things on contact, and improves the documentation in two places (CLI usage and ```local.py```:```scan()```) to clearly state this requirement.

I'm not convinced this is the correct way to introduce this logic. If it exists elsewhere, maybe this should import that somehow and use it instead (repairing if necessary).

Tested to reject only the expected inputs, but requesting review of my entire approach at all. Closes #192 if merged.

### CODE REVIEW

**Comment by marumari** [comment]
I think it might be better to use urlparse(hostname).  If there's a port, stuff it into the port argument and warn them.  Similarly, ignore https://, but warn them that you're scanning 80 and 443.

Any objections?

**Comment by floatingatoll** [comment]
I don't think we should try very hard to parse the arguments here beyond refusing non-hostnames.

If they provide HOST:PORT, then we can't really determine what they meant: did they mean http://PORT or https://PORT? What's the other corresponding PORT that they didn't provide? Would they have to say HOST:PORT,PORT?

If they say http://HOST, should we or should we not scan https://? Allowing them to specify a protocol permits people to guess at the behavior that will occur, and while a warning is theoretically sound, I don't think they should be permitted to reach that stage.

After the discussion in #191 I think that we should simply reject the hostname argument if it's not a hostname, and clearly state that http+https port need to be set, and not offer any opportunity for users to become more confused about which ports and protocols we'll scan.

**Comment by floatingatoll** [comment]
(I was planning to open an issue referencing #191 about "require both http_port and https_port to be provided, if either is provided", along those lines. Haven't yet, pending discussion.)

File: httpobs/scripts/httpobs-local-scan [hunk 1]
```diff
@@ -54,6 +54,23 @@ if __name__ == "__main__":
...
+    #   ... '://', then it has a protocol (http:// or https://).
+    #   ... ']:', then it's a bracketed ipv6 address with a port number.
+    #   ... one colon, then it almost certainly has a port number.
+    # otherwise, it either has:
+    #   ... zero colons: a probably-normal hostname with no port numbers.
+    #   ... two or more colons, and no other markers of port numbers: ipv6 address.
+    #
+    # the above logic avoids regular expressions in favor of code comments
+    # and the following conditional:
+    if '://' in args['hostname'] or ']:' in args['hostname'] or args['hostname'].count(':') == 1:
```
Comment (marumari) [inline httpobs/scripts/httpobs-local-scan]: I try not to be too hacky when I can avoid it.  You can probably just stuff it into urlparse and be sure that it has no scheme (ParseResult.scheme).

For example, this should work:
```python
urlparse(args['hostname']).scheme or urlparse('http://' + args['hostname']).port
```

While also being a lot more self-documenting.  :)

**Comment by floatingatoll** [comment]
Would the valid_hostname() function be the correct path here? Since
protocol port aren't valid hostname - but maybe I misunderstood it's use.

File: httpobs/scripts/httpobs-local-scan (hunk 1 above)
Comment (floatingatoll) [reply httpobs/scripts/httpobs-local-scan]: Got it. Updated the code and it works as intended. Repushed, passes flake+nose.
